codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.codec.http2;
+
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToMessageCodec;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.DefaultLastHttpContent;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.List;
+
+/**
+ * This is a server-side adapter so that an http2 codec can be downgraded to
+ * appear as if it's speaking http/1.1.
+ *
+ * In particular, this handler converts from {@link Http2StreamFrame} to {@link
+ * HttpObject}, and back.  For simplicity, it converts to chunked encoding
+ * unless the entire stream is a single header.
+ */
+public class Http2ServerDowngrader extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
+
+    @Override
+    public boolean acceptInboundMessage(Object msg) throws Exception {
+        return (msg instanceof Http2HeadersFrame) || (msg instanceof Http2DataFrame);
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, Http2StreamFrame frame, List<Object> out) throws Exception {
+        if (frame instanceof Http2HeadersFrame) {
+            int id = 0; // not really the id
+            Http2HeadersFrame headersFrame = (Http2HeadersFrame) frame;
+            Http2Headers headers = headersFrame.headers();
+
+            if (headersFrame.isEndStream()) {
+                if (headers.method() == null) {
```

Comment:
> The Http2Stream allows access to the State.

But that state is 1) not directly usable because it has already been modified in response to the frames, 2) this logic using the state may be delayed, even on the same thread and 3) this logic using the state may be run from a different thread from where the state is maintained/changed.

For point #1, consider these new streams:

IDLE: (isn't actually tracked yet; it will [never](https://github.com/netty/netty/blob/bd6040a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L311) show up as IDLE)
OPEN: Headers (:method=get) StreamId=3 EOS=0 # These are headers
OPEN: Headers () StreamId=3 EOS=1 # These are trailers
HALF_CLOSED_REMOTE: (yes, it happens [afterward](https://github.com/netty/netty/blob/83c349f/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L324))

IDLE: (isn't actually tracked yet; it won't show up as IDLE)
HALF_CLOSED_REMOTE: Headers (:method=get) StreamId=5 EOS=1 # These are headers
(yes, this state happens [immediately](https://github.com/netty/netty/blob/bd6040a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L315))

So the State is _never_ useful for this distinction today. That could possibly be fixed, but the child channel has its own queue. The application could simply not read from the child channel immediately and the state will be HALF_CLOSED_REMOTE in both cases. When the child channel is using a different event loop, we'd expect HALF_CLOSED_REMOTE in both cases.

> This is likely a separate issue but where do we enforce that these "special" headers can only exist "exactly once" before the "data" (unless its a connect request)?

It looks like we should be prohibiting that (still taking into account 1xx):

> Pseudo-header fields are only valid in the context in which they are defined. Pseudo-header fields defined for requests MUST NOT appear in responses; pseudo-header fields defined for responses MUST NOT appear in requests. Pseudo-header fields MUST NOT appear in trailers. Endpoints MUST treat a request or response that contains undefined or invalid pseudo-header fields as malformed (Section 8.1.2.6).